File: impl/ocaml/src/ast.ml
(* The CSS AST (SPEC §4.1.1). Containers hold their children in a mutable
   reference so that walks can splice replacements in place. *)

module SMap = Map.Make (String)

(* Context metadata such as base, reference, theme, source, and sourceBase.
   Boolean flags are stored as "true". *)
type ctx = string SMap.t

type node =
  | Rule of rule
  | At_rule of at_rule
  | Declaration of declaration
  | Comment of comment
  | Context of context
  | At_root of at_root

and nodes = node list ref
and rule = { mutable selector : string; rule_nodes : nodes }
and at_rule = { mutable name : string; mutable params : string; at_nodes : nodes }

(* A declaration with [no_value] is never printed. *)
and declaration = {
  mutable property : string;
  mutable value : string;
  mutable important : bool;
  no_value : bool;
}

and comment = { comment : string }
and context = { context : ctx; ctx_nodes : nodes }
and at_root = { root_nodes : nodes }

let ctx_empty : ctx = SMap.empty
let ctx_of_list l : ctx = List.fold_left (fun m (k, v) -> SMap.add k v m) SMap.empty l
let ctx_get (c : ctx) key = SMap.find_opt key c
let ctx_bool (c : ctx) key = SMap.find_opt key c = Some "true"

(* Returns [c] with [other]'s entries applied. *)
let ctx_merge (c : ctx) (other : ctx) : ctx = SMap.union (fun _ _ v -> Some v) c other

let style_rule ?(nodes = []) selector = Rule { selector; rule_nodes = ref nodes }
let at_rule ?(nodes = []) name params = At_rule { name; params; at_nodes = ref nodes }

let is_space c = c = ' ' || c = '\t' || c = '\n' || c = '\r' || c = '\012'

(* Splits `@media (x)` into name and params. *)
let parse_at_rule_prelude ?(nodes = []) prelude =
  let text = String.trim prelude in
  let n = String.length text in
  let rec find i =
    if i >= n then at_rule ~nodes text ""
    else
      let c = text.[i] in
      if is_space c || c = '(' then
        at_rule ~nodes (String.sub text 0 i) (String.trim (String.sub text i (n - i)))
      else find (i + 1)
  in
  find 1

(* Creates an at-rule when [selector] starts with `@` and a style rule otherwise. *)
let rule ?(nodes = []) selector =
  if String.length selector > 0 && selector.[0] = '@' then parse_at_rule_prelude ~nodes selector
  else style_rule ~nodes selector

let decl ?(important = false) property value =
  Declaration { property; value; important; no_value = false }

let important_decl property value = decl ~important:true property value
let no_value_decl property = Declaration { property; value = ""; important = false; no_value = true }
let comment value = Comment { comment = value }
let context ?(nodes = []) c = Context { context = c; ctx_nodes = ref nodes }
let at_root nodes = At_root { root_nodes = ref nodes }

(* Returns the child list of a container node. *)
let children = function
  | Rule r -> Some r.rule_nodes
  | At_rule a -> Some a.at_nodes
  | Context c -> Some c.ctx_nodes
  | At_root r -> Some r.root_nodes
  | Declaration _ | Comment _ -> None

let rec clone_node = function
  | Rule r -> Rule { selector = r.selector; rule_nodes = ref (clone_nodes !(r.rule_nodes)) }
  | At_rule a -> At_rule { name = a.name; params = a.params; at_nodes = ref (clone_nodes !(a.at_nodes)) }
  | Declaration d -> Declaration { d with property = d.property }
  | Comment c -> Comment { comment = c.comment }
  | Context c -> Context { context = c.context; ctx_nodes = ref (clone_nodes !(c.ctx_nodes)) }
  | At_root r -> At_root { root_nodes = ref (clone_nodes !(r.root_nodes)) }

and clone_nodes nodes = List.map clone_node nodes

type action = Continue | Skip | Stop

(* Passed to a walk visitor. *)
type utils = {
  parent : node option;
  (* The merged context of every enclosing context node. *)
  wctx : ctx;
  (* The ancestors, outermost first. *)
  path : node list;
  mutable replacement : node list option;
}

(* Replaces the current node with [nodes], which are then walked in turn. *)
let replace_with u nodes = u.replacement <- Some nodes

type visitor = node -> utils -> action

(* Walks the tree depth first, parents before children. *)
let rec walk_from (nodes : nodes) (visit : visitor) parent ctx path =
  let rec loop before rest =
    match rest with
    | [] -> Continue
    | node :: tail -> (
        let node_ctx = match node with Context c -> ctx_merge ctx c.context | _ -> ctx in
        let u = { parent; wctx = node_ctx; path; replacement = None } in
        let action = visit node u in
        (match u.replacement with
        | Some repl -> nodes := List.rev_append before (repl @ tail)
        | None -> ());
        if action = Stop then Stop
        else
          match u.replacement with
          | Some repl -> loop before (repl @ tail)
          | None ->
              if action = Skip then loop (node :: before) tail
              else
                let stopped =
                  match children node with
                  | Some ch -> walk_from ch visit (Some node) node_ctx (path @ [ node ]) = Stop
                  | None -> false
                in
                if stopped then Stop else loop (node :: before) tail)
  in
  loop [] !nodes

let walk nodes visit = walk_from nodes visit None ctx_empty []

(* Walks the tree depth first, children before parents. *)
let rec walk_depth_from (nodes : nodes) (visit : node -> utils -> unit) parent ctx path =
  let rec loop before rest =
    match rest with
    | [] -> ()
    | node :: tail -> (
        let node_ctx = match node with Context c -> ctx_merge ctx c.context | _ -> ctx in
        (match children node with
        | Some ch -> walk_depth_from ch visit (Some node) node_ctx (path @ [ node ])
        | None -> ());
        let u = { parent; wctx = node_ctx; path; replacement = None } in
        visit node u;
        match u.replacement with
        | Some repl ->
            nodes := List.rev_append before (repl @ tail);
            loop (List.rev_append repl before) tail
        | None -> loop (node :: before) tail)
  in
  loop [] !nodes

let walk_depth nodes visit = walk_depth_from nodes visit None ctx_empty []

File: impl/go/ast.go
// Package twill implements the Twill utility-first CSS compiler.
package twill

import "strings"

// Node is an element of the AST (SPEC §4.1.1).
type Node interface {
	isNode()
}

// Rule is a style rule.
type Rule struct {
	Selector string
	Nodes    []Node
}

// AtRule is an at-rule. Name includes the leading `@`; Nodes is empty for
// statement at-rules.
type AtRule struct {
	Name   string
	Params string
	Nodes  []Node
}

// Declaration is a property declaration. A declaration with NoValue is
// never printed.
type Declaration struct {
	Property  string
	Value     string
	Important bool
	NoValue   bool
}

// Comment is a kept `/*!` comment; Value is the text between the markers.
type Comment struct {
	Value string
}

// ContextMap carries metadata such as base, reference, theme, source, and
// sourceBase. Boolean flags are stored as "true".
type ContextMap map[string]string

// Bool reports whether the flag key is set.
func (c ContextMap) Bool(key string) bool { return c[key] == "true" }

// Merged returns a copy of c with other's entries applied.
func (c ContextMap) Merged(other ContextMap) ContextMap {
	m := make(ContextMap, len(c)+len(other))
	for k, v := range c {
		m[k] = v
	}
	for k, v := range other {
		m[k] = v
	}
	return m
}

// Context is never printed; it carries metadata to its subtree.
type Context struct {
	Context ContextMap
	Nodes   []Node
}

// AtRoot is never printed in place; its children are hoisted to the end of
// the document during optimization.
type AtRoot struct {
	Nodes []Node
}

func (*Rule) isNode()        {}
func (*AtRule) isNode()      {}
func (*Declaration) isNode() {}
func (*Comment) isNode()     {}
func (*Context) isNode()     {}
func (*AtRoot) isNode()      {}

// StyleRule creates a style rule.
func StyleRule(selector string, nodes ...Node) *Rule {
	if nodes == nil {
		nodes = []Node{}
	}
	return &Rule{Selector: selector, Nodes: nodes}
}

// NewAtRule creates an at-rule.
func NewAtRule(name, params string, nodes ...Node) *AtRule {
	if nodes == nil {
		nodes = []Node{}
	}
	return &AtRule{Name: name, Params: params, Nodes: nodes}
}

// NewRule creates an at-rule when selector starts with `@` and a style rule
// otherwise.
func NewRule(selector string, nodes ...Node) Node {
	if strings.HasPrefix(selector, "@") {
		return ParseAtRulePrelude(selector, nodes...)
	}
	return StyleRule(selector, nodes...)
}

// ParseAtRulePrelude splits `@media (x)` into name and params.
func ParseAtRulePrelude(prelude string, nodes ...Node) *AtRule {
	text := strings.TrimSpace(prelude)
	for i := 1; i < len(text); i++ {
		c := text[i]
		if c == ' ' || c == '\t' || c == '\n' || c == '\f' || c == '\r' || c == '(' {
			return NewAtRule(text[:i], strings.TrimSpace(text[i:]), nodes...)
		}
	}
	return NewAtRule(text, "", nodes...)
}

// Decl creates a declaration.
func Decl(property, value string) *Declaration {
	return &Declaration{Property: property, Value: value}
}

// ImportantDecl creates an important declaration.
func ImportantDecl(property, value string) *Declaration {
	return &Declaration{Property: property, Value: value, Important: true}
}

// NewComment creates a comment node.
func NewComment(value string) *Comment { return &Comment{Value: value} }

// NewContext creates a context node.
func NewContext(ctx ContextMap, nodes ...Node) *Context {
	if nodes == nil {
		nodes = []Node{}
	}
	if ctx == nil {
		ctx = ContextMap{}
	}
	return &Context{Context: ctx, Nodes: nodes}
}

// NewAtRoot creates an at-root node.
func NewAtRoot(nodes ...Node) *AtRoot {
	if nodes == nil {
		nodes = []Node{}
	}
	return &AtRoot{Nodes: nodes}
}

// Children returns a pointer to the child list of a parent node, or nil.
func Children(n Node) *[]Node {
	switch n := n.(type) {
	case *Rule:
		return &n.Nodes
	case *AtRule:
		return &n.Nodes
	case *Context:
		return &n.Nodes
	case *AtRoot:
		return &n.Nodes
	}
	return nil
}

// CloneNodes deep-clones a node list.
func CloneNodes(nodes []Node) []Node {
	out := make([]Node, len(nodes))
	for i, n := range nodes {
		out[i] = CloneNode(n)
	}
	return out
}

// CloneNode deep-clones a node.
func CloneNode(n Node) Node {
	switch n := n.(type) {
	case *Rule:
		return &Rule{Selector: n.Selector, Nodes: CloneNodes(n.Nodes)}
	case *AtRule:
		return &AtRule{Name: n.Name, Params: n.Params, Nodes: CloneNodes(n.Nodes)}
	case *Declaration:
		d := *n
		return &d
	case *Comment:
		c := *n
		return &c
	case *Context:
		return &Context{Context: n.Context.Merged(nil), Nodes: CloneNodes(n.Nodes)}
	case *AtRoot:
		return &AtRoot{Nodes: CloneNodes(n.Nodes)}
	}
	return n
}

// WalkAction controls a walk.
type WalkAction int

const (
	// Continue descends into children.
	Continue WalkAction = iota
	// Skip does not descend into this node's children.
	Skip
	// Stop ends the whole walk.
	Stop
)

// WalkUtils is passed to a walk visitor.
type WalkUtils struct {
	// Parent is the parent node, or nil at the top level.
	Parent Node
	// Context is the merged context of every enclosing context node.
	Context ContextMap
	// Path holds the ancestors, outermost first.
	Path []Node

	replaced    bool
	replacement []Node
}

// ReplaceWith replaces the current node with the given nodes, which are
// then walked in turn.
func (u *WalkUtils) ReplaceWith(nodes ...Node) {
	u.replaced = true
	u.replacement = nodes
	if u.replacement == nil {
		u.replacement = []Node{}
	}
}

// Visitor visits a node during a walk.
type Visitor func(n Node, u *WalkUtils) WalkAction

// Walk walks the tree depth first, parents before children. Replacement
// nodes are visited in turn.
func Walk(nodes *[]Node, visit Visitor) WalkAction {
	return walk(nodes, visit, nil, ContextMap{}, nil)
}

// WalkFrom walks with an explicit parent, context, and path.
func WalkFrom(nodes *[]Node, visit Visitor, parent Node, ctx ContextMap, path []Node) WalkAction {
	return walk(nodes, visit, parent, ctx, path)
}

func walk(nodes *[]Node, visit Visitor, parent Node, ctx ContextMap, path []Node) WalkAction {
	for i := 0; i < len(*nodes); i++ {
		node := (*nodes)[i]
		nodeCtx := ctx
		if c, ok := node.(*Context); ok {
			nodeCtx = ctx.Merged(c.Context)
		}
		u := &WalkUtils{Parent: parent, Context: nodeCtx, Path: path}
		action := visit(node, u)
		if u.replaced {
			rest := append([]Node{}, (*nodes)[i+1:]...)
			*nodes = append(append((*nodes)[:i], u.replacement...), rest...)
		}
		if action == Stop {
			return Stop
		}
		if u.replaced {
			i--
			continue
		}
		if action == Skip {
			continue
		}
		if children := Children(node); children != nil {
			if walk(children, visit, node, nodeCtx, append(path, node)) == Stop {
				return Stop
			}
		}
	}
	return Continue
}

// WalkDepth walks the tree depth first, children before parents.
func WalkDepth(nodes *[]Node, visit func(n Node, u *WalkUtils)) {
	walkDepth(nodes, visit, nil, ContextMap{}, nil)
}

func walkDepth(nodes *[]Node, visit func(n Node, u *WalkUtils), parent Node, ctx ContextMap, path []Node) {
	for i := 0; i < len(*nodes); i++ {
		node := (*nodes)[i]
		nodeCtx := ctx
		if c, ok := node.(*Context); ok {
			nodeCtx = ctx.Merged(c.Context)
		}
		if children := Children(node); children != nil {
			walkDepth(children, visit, node, nodeCtx, append(path, node))
		}
		u := &WalkUtils{Parent: parent, Context: nodeCtx, Path: path}
		visit(node, u)
		if u.replaced {
			rest := append([]Node{}, (*nodes)[i+1:]...)
			*nodes = append(append((*nodes)[:i], u.replacement...), rest...)
			i += len(u.replacement) - 1
		}
	}
}
